pool.go

```diff
@@ -24,28 +24,85 @@ import (
 )
 
 type modulePool[T interfaces.Signature] struct {
-	pool sync.Pool
-	new  func() (*module[T], error)
+	pool    sync.Pool
+	maxSize uint32
+	new     func(*module[T]) (*module[T], error)
+	close   func(*module[T])
+	ch      chan *module[T] // for buffered channel impl
 }
 
-func newModulePool[T interfaces.Signature](ctx context.Context, template *template[T]) *modulePool[T] {
-	return &modulePool[T]{
-		new: func() (*module[T], error) {
-			return newModule[T](ctx, template)
-		},
+func newModulePool[T interfaces.Signature](ctx context.Context, template *template[T], maxSize uint32) *modulePool[T] {
+	if maxSize == 0 {
+		// If size = 0, use a standard sync.Pool + golang's runtime finalizers
+		// to make sure the close function gets called eventually.
+		return &modulePool[T]{
+			maxSize: maxSize,
+			new: func(m *module[T]) (*module[T], error) {
+				m.SetFinalizer()
+				return newModule[T](ctx, template)
+			},
+			close: func(m *module[T]) {
+				m.Close()
+			},
+		}
 	}
+
+	if maxSize > 0 {
+		// if size > 0 then we use buffered channel implementation.
+		return &modulePool[T]{
+			maxSize: maxSize,
+			new: func(m *module[T]) (*module[T], error) {
+				return newModule[T](ctx, template)
+			},
+			close: func(m *module[T]) {
+				m.Close()
+			},
+			ch: make(chan *module[T], maxSize),
+		}
+	}
+
+	return nil
 }
 
 func (p *modulePool[T]) Put(m *module[T]) {
-	if m != nil {
+	if m == nil {
+		return
+	}
+
+	if p.maxSize == 0 {
 		p.pool.Put(m)
+	} else {
+		select {
+		case p.ch <- m:
+		default:
+			// Channel is full, call the close function
+			p.close(m)
+		}
 	}
```

Comment:
Should be safe to always push to the channel and default to the pool. If the channel is unbuffered (which is the case when `maxSize = 0`)  then you'd automatically drop back to the pool. I personally would always let the pool do the module deallocation when it wants to and never manually call `p.close` the whole point of the pool (imo) is that allocated modules can be shared and I think there are some wins there even when `maxSize > 0`.

Both approaches are pretty close to equivalent in practice, this approach (if valid) is probably just simpler to reason about.

```suggestion
	select {
        case p.ch <- m:
        default:
	        p.pool.Put(m)
        }
```